site/guides/variant-constructor-pattern.md

```diff
@@ -0,0 +1,182 @@
+# Variant construction pattern
+
+If you look at the config workflows, you will notice that creating variant
+of a package is a very common and frequent operation, so reducing the steps
+required to create a variant can have significant benefits for the
+package consumers. In this guide, we will look at some techniques
+that a package author can use to enable automatic variant construction of a package.
+
+## Types of packages
+
+kpt packages comes in two flavors:  `abstract package` and
+`deployable package instance`. An `abstract` package is a reususable package that
+is used to create deployable instances that can be deployed to a
+kubernetes cluster. In programming language terms, you can think of an `abstract`
+packages as the class and `deployable instance` as the instances of the class.
+`deployable` instances of package are also referred to as `variant` of the package.
+
+Figure below shows a `package catalog` on the left that has `abstract` packages
+and `deployable instances` on the right. A good pattern is to keep the abstract
+packages and instance packages in separate repos and typically
+`deployable instances` repo will be setup to auto deploy to a kubernetes cluster
+using gitops tools such as `config-sync`, `fluxcd`, `argocd`.
+
+![variant constructor pkg repo diagram](/static/images/variant-constructor-pkg-repo-diagram.png)
+
+Resources in an `abstract` package have placeholder values that needs to be
+substituated with actual values to make them ready for deployment.
+For example, name of the namespace resource below has `example` as a placeholder
+value. This is a part of the `abstract package`.
+
+```yaml
+apiVersion:v1
+kind: Namespace
+metadata:
+  name: example # <-- this is a placeholder value
+```
+
+## Customizing identity of resources
+
+A kpt package contains kubernetes resources. Whenever you are creating a
+variant of the package, first step is to ensure unique identity of the
+resources in that variant. For example, if the abstract package contains a
+`namespace` resource, then the variant package should contain namespace resource
+corresponding to that variant.
+
+In a kubernetes cluster, resources are identified by their group, version, kind,
+namespace and name (aka GVKNN). If resource is cluster scoped, then the
```

Comment:
```suggestion
namespace and name (sometimes referred to as GVKNN). If resource is cluster scoped, then the
```

(in case there are people around the world not familiar with aka)